### DISCUSSION TITLE
Better type errors (rerun type checking on error)

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#8509] [Ideas]
Type errors are often bad for two reasons
- unresolved constraints (will be fixed in the new solver by failing slowly)
- backwards information flow

Backwards information flow is hard to correct in a performant way, but what if we just rerun type checking for functions with errors? This way errors would be more intuitive while keeping the happy case performant

**Comment by magnus-madsen** [comment]
Why would rerunning the constraint-based solver give different results? I think it should be idempotent.

**Reply by JonathanStarup** [reply]
Rerunning without index-based sorting gives very different results - that's not a bad thing. it's an optimization you can turn of.

Maybe it should be idempotent, but while it isn't, and having the flag is so cheap, i think its wasted not to use

**Comment by JonathanStarup** [comment]
Confirmed to work very well for errors in  #8515 but can blow up inference complexity without the ordering.

**Comment by magnus-madsen** [comment]
Do you have an example where this is needed with the new solver?

**Reply by JonathanStarup** [reply]
Yes? Most programs have better errors with this change and nothing's changed with the new solver. The example from the PR still stands
```scala
def example(rc: Region[r]): Int32 =
    let _r = Ref.fresh(rc, 42);
    42
```
```
>> Unable to unify the effect formulas: 'Pure' and 'r'.

2 |     let _r = Ref.fresh(rc, 42);
                           ^^
                           mismatched effect formulas.
```
Rerunning with in-order constraints give the intuitive error that the effect of the function is wrong
```
>> Expected type: 'Pure' but found type: 'r'.

1 |> def example(rc: Region[r]): Int32 =
2 |>     let _r = Ref.fresh(rc, 42);
3 |>     42
```
